File: SDBC/Python/flowplot.py
# coding=utf-8
# copyright(c) 2021 Central South University. all rights reserved.
# jianbo.tang@csu.edu.cn
# 2024/10/15

import matplotlib
import numpy as np
import matplotlib.pyplot as plt
plt.style.use('ggplot')
plt.rcParams['font.family'] = 'Arial'
# plt.yticks(fontproperties = 'Times New Roman', size = 14)
# plt.xticks(fontproperties = 'Times New Roman', size = 14)


def flowplot(flows, IDX=None, linewidth=2):
    if IDX is None:
        IDX = np.ones(flows.shape[0], dtype=int)

    n = flows.shape[0]
    if len(IDX) < n:
        IX = np.zeros(n, dtype=int)
        IX[IDX] = 1
        IDX = IX

    IX = np.unique(IDX[IDX > 0])
    n = len(IX)
    cluster_colors = 'b'
    if n > 1:
        cluster_colors = np.random.rand(n, 3)
    noise_colors = [0.7, 0.7, 0.7]

    X, Y = flow_to_sequence(flows, IDX < 1)
    plt.plot(X, Y, '-', color=noise_colors, linewidth=linewidth)

    for i in range(len(IX)):
        X, Y = flow_to_sequence(flows, IDX == IX[i])
        plt.plot(X, Y, '-', color=cluster_colors[i] if n > 1 else cluster_colors, linewidth=linewidth)

    plt.xlabel('X', fontsize=14, fontweight='bold')
    plt.ylabel('Y', fontsize=14, fontweight='bold')


def hold(mode='on'):
    if mode == 'on':
        plt.ion()
    else:
        plt.ioff()


def draw_now(fig=None, sleep_sec=0.001):
    if fig is None:
        fig = plt.gcf()
    fig.canvas.draw()
    fig.canvas.flush_events()
    plt.pause(sleep_sec)


def flow_to_sequence(flows, flow_ids):
    X = list()
    Y = list()
    if np.any(flow_ids):
        n = len(flow_ids)
        if n == flows.shape[0]:
            n = np.sum(flow_ids)
        X = np.full(3 * n, np.nan)
        Y = np.full(3 * n, np.nan)
        X[::3] = flows[flow_ids, 0]
        X[1::3] = flows[flow_ids, 2]
        Y[::3] = flows[flow_ids, 1]
        Y[1::3] = flows[flow_ids, 3]
    return X, Y


# main
if __name__ == '__main__':
    pass

File: SDBC/Octave/flowplot.m
function flowplot(flows, IDX, linewidth)

%  Readme:
%    Code of the paper "Statistical and density-based clustering of 
%    geographical flows for crowd movement patterns recognition, Applied
%    Soft Computing, 2024, 163: 111912"
%    DOI: https://doi.org/10.1016/j.asoc.2024.111912
%  If you use this code, please cite the above paper, Thanks.
%  Tang Jianbo, CSU
%

if nargin<2
    IDX = ones(size(flows,1), 1);
end
if nargin<3
    linewidth = 2;
end

n = size(flows, 1);
if length(IDX) < n
    IX = zeros(n,1);
    IX(IDX) = 1;
    IDX = IX;
    clear IX;
end

hold on;
% set(gcf, 'position', [544.1111  163.2222  772.0000  691.5556]);
IX = unique(IDX(IDX>0));
n = length(IX);
cluster_colors = 'b';
if n==1
    cluster_colors = [1 0 0]; %colors = {'k','r','b','g','y','c','m','p'};
elseif n>1
    cluster_colors = rand(n, 3); %colors = {'k','r','b','g','y','c','m','p'};
end
noise_colors = [0.7 0.7 0.7];

[X, Y] = flow_to_sequence(flows, IDX<1);
if ~isempty(X)
    plot(X, Y, '-', 'color', noise_colors, 'linewidth', linewidth);
end

for i=1:length(IX)
    [X, Y] = flow_to_sequence(flows, IDX==IX(i));
    if ~isempty(X)
        plot(X, Y, '-', 'color', cluster_colors(i,:), 'linewidth', linewidth);
    end
end
axis('equal');
xlabel('X');
ylabel('Y');
set(gca, 'fontsize', 14, 'fontweight', 'bold');
drawnow;
end % flowplot


function [X, Y] = flow_to_sequence(flows, label)
if any(label)
    n = length(label);
    if n==size(flows, 1)
        n = sum(label);
    end
    X = NaN*ones(3*n, 1);
    Y = NaN*ones(3*n, 1);
    X(1:3:end) = flows(label,1);
    X(2:3:end) = flows(label,3);
    Y(1:3:end) = flows(label,2);
    Y(2:3:end) = flows(label,4);
else
    X = [];
    Y = [];
end
end % func


